### PR TITLE
[github actions] Automate Flutter Chery Picks

### PR DESCRIPTION
**Pull request by XilaiZhang**
**design doc: go/easy-cp**
umbrella design doc: go/flutter-actions
umbrella bug: https://github.com/flutter/flutter/issues/139604 

**Sample Results**:
1. If cherry pick succeeds, a pull request with cherry pick template and label is created.
https://github.com/flutter/flutter/pull/139523
2. If cherry pick fails, a comment is added under the original Pull Request.
https://github.com/flutter/cocoon/pull/3305

In tests, [7d9010](https://github.com/flutter/flutter/commit/7d9010c3578338bc850d66c315bd6086f3d3dc9a) was used to simulate a clean cherry pick, and [cf71a5](https://github.com/flutter/flutter/commit/cf71a55f7d1613dd1795c68cd799c18900caa39a) was used to simulate a merge conflict during cherry pick.

**Implementation Details:**
1. triggered when 'cp: beta' or 'cp:stable' label is added to the original PR
4. parses release channel and gets release candidate branch name
5. get commit sha from event payload
6. checks out framework repo and revision history
7. Attempt a cherry pick without any resolution strategy
8. If cp is successful, uses a PR template to open a pull request. 
9. If cp isn't successful, leave a comment on the original PR.

**PR template**
Since PR template doesn't support web form, the cherry pick PR template we used is https://github.com/XilaiZhang/miscellaneous-side-project/blob/master/.github/workflows/template/cherrypick.md, which is adapted from the [web form cherry pick issue template](https://cs.opensource.google/flutter/flutter/+/master:.github/ISSUE_TEMPLATE/7_cherry_pick.yml).
This PR template should be reviewed and added to be under the .github path in framework repository. 

**Decisions Taken**
1. place of source code
I put the source code under the framework repository. Acknowledges the risk of duplication of code, in favor of not needing to package and publish our own actions to market place.
more details in [open discussions section of design doc](https://docs.google.com/document/d/1EUyJ9NCAltxJq3P3pIzIRVF2ArPUO4cj_eqIUvzl8cw/edit#heading=h.e3zrsz9gwxmp)

2. Resolution strategy
No resolution strategy is applied at all during the cherry pick process. [more details](https://docs.google.com/document/d/1EUyJ9NCAltxJq3P3pIzIRVF2ArPUO4cj_eqIUvzl8cw/edit#heading=h.3za1b9qx6pi2)

**Future Work**
A PAT token is needed for authorization to create branch and add comment. We would need to create a new bot and update its credentials in the secrets section of flutter/flutter repository.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: .github/workflows/easy-cp.yml
Comment (CaseyHillers) [inline .github/workflows/easy-cp.yml:27]: Is it possible to store this in memory instead of writing a file?

```sh
RELEASE_BRANCH=$(curl https://raw.githubusercontent.com/flutter/flutter/$CHANNEL/bin/internal/release-candidate-branch.version)
```

File: .github/workflows/easy-cp.yml
Comment (CaseyHillers) [inline .github/workflows/easy-cp.yml:24]: Is the double echo unnecessary?

export CHANNEL=$(${{ github.event.label.name }} | cut -d ':' -f 2)

File: .github/workflows/easy-cp.yml
Comment (CaseyHillers) [inline .github/workflows/easy-cp.yml:53]: Should this be implemented now?

File: .github/workflows/easy-cp.yml
Comment (CaseyHillers) [inline .github/workflows/easy-cp.yml:40]: Why is depth 0? It would help to add a comment here explaining why, otherwise people have to check below

File: .github/workflows/easy-cp.yml
Comment (CaseyHillers) [inline .github/workflows/easy-cp.yml:52]: Should `git status` be removed?

File: .github/workflows/easy-cp.yml
Comment (CaseyHillers) [inline .github/workflows/easy-cp.yml:50]: Should we include the channel in the branch? I expect there's a good use case where people will want to open CPs at the same time to both beta and stable

File: .github/workflows/easy-cp.yml
Comment (CaseyHillers) [inline .github/workflows/easy-cp.yml:65]: Instead of using a PAT, is the plan to follow https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs to grant write access to this job's GITHUB_TOKEN?

File: .github/workflows/easy-cp.yml
Comment (CaseyHillers) [inline .github/workflows/easy-cp.yml:66]: Can we re-use the original commit message?

File: .github/workflows/easy-cp.yml
Comment (CaseyHillers) [inline .github/workflows/easy-cp.yml:55]: Is this needed if below we use "body-path" to use the template?

File: .github/workflows/easy-cp.yml
Comment (CaseyHillers) [inline .github/workflows/easy-cp.yml:78]: We'll need to update docs on how to create these CPs. Seems like we'll need to move some of the documentation that exists in the CP issue template into the wiki, like where to find the branch, and to use the CP template as the body.

```suggestion
          FAILURE_MSG+="Failed to create CP due to merge conflicts. You will need to create the PR manually. See https://github.com/flutter/flutter/wiki/Flutter-Cherrypick-Process"
```

File: .github/workflows/easy-cp.yml
Comment (CaseyHillers) [inline .github/workflows/easy-cp.yml:76]: nit: Remove the first two lines of the error message. Since this is being posted by a bot account, most people will already get this from the context.

**Review by CaseyHillers — COMMENTED** [review]
1. Can we create a flutter.dev/go/ link so anyone can access the design doc?
2. Can we link to the tracking bug?

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml]: Sure, removed.

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml]: Updated failure message.

Umm I think the part on where to find the branch information might not be needed, since we can retrieve it from the github action payload and do it for the user.

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml:54]: Yeah i added a TODO here but we can only remove this part after https://github.com/flutter/flutter/pull/139590 lands (or when Kevin is back).

Without an existing template in flutter/flutter, we will have to download the template from somewhere and include it in the opened pull request.

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml:64]: My understanding is that there isn't an original commit message, because the `message` associated with a merged commit sha (cp commit sha), is the PR description of the merged PR.

e.g. the merged commit sha is af61c6c73fb1057633c84bde5b183f8fc3b0673a of https://github.com/flutter/flutter/pull/139348. running the query `gh search commits --repo flutter/flutter --hash e5762ba --json commit` gives a commit message that is several hundred chars (the pr description)

e.g. in the cherry pick pr https://github.com/flutter/flutter/pull/139594. The commit message associated with the cherry picked commit is the original PR title.

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml]: Yeah we have the line `permissions: write-all`.

Good point! did some tests and I think it works on the same repo where the action is installed. e.g. the github.token of flutter/flutter cannot comment on flutter/engine. but should work on flutter/flutter after we merge this PR.

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml]: oh the branch name here actually has to be the same as $RELEASE_BRANCH. updated.

i.e., it has to be `git checkout -b $RELEASE_BRANCH --track origin/$RELEASE_BRANCH` cause the market place action does a hard reset on this name. 
The marketplace action has a step called `Create or update the pull request branch` that always does `/usr/bin/git reset --hard origin/$BRANCH_NAME`. if we use a fancy branch name other than $RELEASE_BRANCH, the market place action can't find it from remote and will throw an error.

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml]: Sure, removed.

Yeah previously added it as a sanity check for myself.

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml:40]: Sure, added a comment.

yeah briefly mentioned in the PR description and it checks out history on the branch.

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml:52]: Sure, created https://github.com/flutter/flutter/pull/139590.

The template has to be in markdown format cause the original web form issue template (https://github.com/flutter/flutter/blob/master/.github/ISSUE_TEMPLATE/7_cherry_pick.yml) isn't supported in PR description.

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml:24]: Tested and my understanding is that both the inner and outer echos are needed.

My understanding is that the inner `echo` is needed to pipe it to the next expression.
without the inner echo, `$word | cut -d ':' -f 2` would throw command not found error on $word
with the echo, `echo $word | cut -d ':' -f 2` would pipe the output of `echo $word` into the next expression

The outer `echo` is needed due to the unique way of storing environment variables in github actions.
without the outer echo and using `export $var`, the variable is not stored in github actions. [failed workflow](https://github.com/XilaiZhang/miscellaneous-side-project/actions/runs/7106918754/job/19347336352)
with the outer echo, Echo it into GITHUB_ENV would prevserve it. (https://stackoverflow.com/questions/57968497/how-do-i-set-an-env-var-with-a-bash-expression-in-github-actions)

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml]: Sure, changed from file to variable

**Comment by XilaiZhang** [comment]
> 1. Can we create a flutter.dev/go/ link so anyone can access the design doc?
> 2. Can we link to the tracking bug?

Sure, created a tracking bug at https://github.com/flutter/flutter/issues/139604.
Umm looked around on flutter.dev but doesn't seem to find a place that archives design docs

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml:24]: Umm yeah I haven't found a source that includes all the supported syntax. Currently my learnings come from failed workflows and stackoverflow.

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml:52]: yeah I implemented separately cause I thought if the template gets merged into flutter/flutter first, then we can directly use it in github actions. otherwise even if they are in the same PR, the github action still has to download the template from somewhere else because the template doesn't exist on framework yet.

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml:39]: Yeah so `fetch-depth` means the number of commits to fetch, and `fetch-depth: x` means fetching the last x commits on branch `ref`.

For the cherry pick commit, its unclear where the commit is on the master branch. e.g. the cp commit could be located at 201 commits before HEAD, and therefore `fetch-depth: 200` would not retrieve the cp commit.

I had to do a `fetch-depth:0` here (0 fetches all commits on `ref`), so that no matter where the cp commit is in the commit history, it is always retrieved and a known object. Even if it is the first commit of the repository. 

By default, fetch-depth is set to 1 which only retrieves the tip commit on branch `ref`.

File: .github/workflows/easy-cp.yml
Comment (CaseyHillers) [reply .github/workflows/easy-cp.yml]: Is this something `branch-suffix` can help with?

File: .github/workflows/easy-cp.yml
Comment (CaseyHillers) [reply .github/workflows/easy-cp.yml]: Do we need the create pull request action? Can we use the gh cli directly?

File: .github/workflows/easy-cp.yml
Comment (CaseyHillers) [reply .github/workflows/easy-cp.yml:54]: With https://github.com/flutter/flutter/pull/139590 ready, can we remove the TODO?

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml]: Good point, updated pushed branch name to support multiple CPs.

Yeah the branch names we checkout here in the container are local to our local checkout, and doesn't affect the branch name pushed to the release branch. later when we open the pull request, the pushed branch will have a different name. updated the name of the feature branch to be `cp-${{ env.CHANNEL }}-${{ env.COMMIT_SHA }}` via the `branch` field of the marketplace action.

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml:53]: Tested around and realized I cannot remove this step because the template isn't on release branches yet.

Currently the template is only on master head and not on release branches. When we are creating the pull request from the release branch, this file path becomes unavailable to us. For now, I still have to download it from the flutter/flutter repo.

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml]: Yeah we can definitely use gh cli directly. 

My first few drafts are all implemented with pure gh/git commands, but I think in previous design/code reviews it was pointed out that I should use market place actions whenever possible. I am good with reimplementing with gh and git commands again. 

However, if the concern here is that an unnecessary commit message will be generated, then I can assure that no duplicate/extra commit message will be produced. See https://github.com/flutter/flutter/pull/139843 as an example where I cp-ed a chillers pr using market place action.

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml]: Updated the comment to `Checkout all history commits on master branch, so that the cp commit is a known object`

I might have misunderstood and I am good with whichever comment that would make this part clearer. If Casey proposes a change here I d be happy to accept it.

File: .github/workflows/easy-cp.yml
Comment (XilaiZhang) [reply .github/workflows/easy-cp.yml]: Sure, updated

**Comment by XilaiZhang** [comment]
> Please attach it to the pull request description above. For cases like this, it's helpful to keep the standard pull request template to make sure the PR adheres to the repos guidelines. (see https://github.com/flutter/flutter/blob/master/.github/PULL_REQUEST_TEMPLATE.md)

Sure, attached

> https://github.com/flutter/flutter/issues/new?assignees=&labels=design+doc&projects=&template=8_design_doc.yml

Oh TIL, created an issue at https://github.com/flutter/flutter/issues/139846

**Comment by CaseyHillers** [comment]
Do we know why [Google testing](https://github.com/flutter/flutter/pull/139524/checks?check_run_id=19468129192) is stuck in the queued state for 2+ days?

File: .github/workflows/easy-cp.yml
Comment (CaseyHillers) [reply .github/workflows/easy-cp.yml:24]: I found https://docs.github.com/en/actions/learn-github-actions/essential-features-of-github-actions which covers the basics

**Review by CaseyHillers — APPROVED** [review]

**Pull request merged**
